File: _programmers/_28_/marathon.js
function solution(participant, completion) {
  const _participant = participant
    .reduce((acc, curr) => {
      if (!(curr in acc)) {
        acc[curr] = 1
      } else {
        acc[curr] += 1
      }
      return acc
    }, {})

  for (let comp of completion) {
    _participant[comp] -= 1
  }
  
  return Object.entries(_participant)
    .filter(e => e[1] > 0)[0][0]
}

console.log(solution(
  ["leo", "kiki", "eden"],
  ["eden", "kiki"]
))

File: _programmers/_28_/Marathon.java
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
// import static java.util.Map.Entry;
// import java.util.stream.Collector;
import java.util.stream.Collectors;

class Solution {
  public String solution(String[] participant, String[] completion) {
    Map<String, Long> map = Arrays
        .stream(participant)
        .collect(Collectors.groupingBy(p -> p, HashMap::new, Collectors.counting()));

    for (String comp: completion) {
      map.put(comp, map.get(comp) - 1);
    }

    return map.entrySet()
      .stream()
      .filter(e -> e.getValue() > 0)
      .findFirst().get().getKey();
  }
}

public class Marathon {
  public static void main(String[] args) {
    System.out.println(
    new Solution()
      .solution(
        new String[] {
          "leo", 
          "kiki", 
          "eden"
        },
        new String[] {
          "eden",
          "kiki" 
        }
      )
    );
  }
}